### ISSUE TITLE
go style errors

### ISSUE THREAD

**Issue by joshrtay** [matthewmueller/poss#1]
this is interesting... I thought about doing something similar, but handling them go style. something like:

``` js
let [resources, err] = yield poss(fetch(token))
if (err) return err
```

**Comment by matthewmueller** [comment]
Oh nice :-) I didn't realize that's how Go did it. 

I'll probably leave it as is for now since i don't use babel everywhere. So I'd end up with stuff like this:

``` js
let res = yield poss(fetch(token))
if (res[1]) return err
```

which isn't nice. Super easy to come up with your own flavor though – it's only 4 lines :-)

**edit:** does node 5 support spreads? it looks like they might.

**Comment by joshrtay** [comment]
ya. it's a terrible syntax without destructing. 

doesn't look like node supports it yet.. someday :)

**Comment by matthewmueller** [comment]
node 6.0.0 supports destructuring :-)

wonder if it makes sense to be an array or an object:

``` js
let { ret, err } = yield poss(fetch(token))
```

any thoughts?

**Comment by joshrtay** [comment]
One thing to keep in mind is that JS supports partial destructuring, while go does not. I think part of the reason that go's error handling convention works is that you are forced to destructure the error. In JS it would be really easy to "drop" an error. 

``` js
let [a] = [1, err]
// or
let {a} = {a: 1, err: err}
```

The only way I can think of to avoid dropping errors would be to do error first array destructuring. So rather than my previous example you would do:

``` js
let [err, resource] = yield poss(fetch(token))
if (err) return err
```

**Comment by matthewmueller** [comment]
great point! i originally was thinking that the partial destructuring was a good/cleaner thing, but i can see it being really easy to shoot yourself in the foot with that. i'll roll what you suggested then :-)

**Comment by ashaffer** [comment]
+1. I'd like to use this, but I can't guarantee my errors will always be `instanceof Error`, so the in-band signaling makes the return value ambiguous.

**Comment by matthewmueller** [comment]
done in 2.x!

**Comment by joshrtay** [comment]
Cool! What was your reasoning behind error second?

**Comment by matthewmueller** [comment]
ah, just looking at how go does it: https://blog.golang.org/error-handling-and-go

the thought for error in front is how node typically does it?

**Comment by joshrtay** [comment]
ya. and that go doesnt support partial restructuring, but javascript does. so the only way to avoid "dropping" the error in javascript is to put it first.

**Comment by matthewmueller** [comment]
ahh okay good call, i'll change it and re-release :-)

**Comment by matthewmueller** [comment]
released in 3.0.0. thanks for the feedback @joshrtay!
